File: problem_7/problem_7.rb
def prime?(num)
  return true if [1, 2].include?(num)
  return false if num.even?

  n = Math.sqrt(num).ceil
  div = nil

  3.upto(n).select(&:odd?).each do |d|
    if num % d == 0
      div = d
      break
    end
  end

  div.nil?
end

n = 3
pl = 2

while pl < 10001 do
  n += 2
  pl += 1 if prime?(n)
end

puts n

File: problem_7/problem_7.clj
(ns project-euler.problem_7)

;;Решение в лоб
(defn- any-divisors?
  [x n]
  (if (= n 1)
    false
    (if (and
          (= (rem x n) 0)
          (not (= x n)))
      true
      (any-divisors? x (dec n)))))

(defn- prime?
  [x]
  (let [n (int (Math/ceil (Math/sqrt x)))]
    (if (or (and
              (even? x)
              (not (= x 2)))
            (= x 1))
      false
      (not (any-divisors? x n)))))

(defn find-nth-prime
  [n]
  (last (take n (filter prime? (iterate inc 1)))))

(prn (find-nth-prime 10001))

;;Решето Эратосфена

